portfolio/src/main/webapp/projects/commentart/scripts/main.js

```diff
@@ -0,0 +1,18 @@
+/** Sets up all the required listeners and creates all necessary elements. */
+const main = () => {
+  const body = document.getElementsByTagName('body')[0];
+  body.onload = fetchAndShowComments;
+}
+
+/** Fetches comments from the server. */
+const fetchAndShowComments = () => {
+  fetch('/read_comments')
+      .then(response => response.json())
+      .then(commentsJson => {
+        const comments = commentsJson.map(Comment.toComment(commentJson));
+        const commentsManager = new CommentsManager(comments);
```

Comment:
Why do we need to create a new CommentsManager for every fetch request?

Reply:
This fetch request is done to load data from the server. It will only be called once the the page reloads. I suppose, I need a manager for each session, don't I? :O

Reply:
First, the name `CommentsManager` is a bit confusing, it seems to be managing something but actually it's only responsible for creating some UI elements. I think that on the front-end side, we don't need so many abstraction for UIs, because they're just UI logic. Front-end should only get data from server, and display accordingly. 

In terms of creating a separate class for Comment, Reply, I don't find them currently being necessary for front-end. Each time we create a separate class, it must be that we need different instantiations of it and we need to keep the reference of it for future use, which I don't find to be super convincing right now. Otherwise this will only increase more memory usage by keeping UI elements in the memory.

Reply:
Hmm... I see where you are coming from. Perhaps the above comment (which I just posted) is relevant here. 

Oh okay! But if I do not maintain classes, how do I ensure that when the user clicks a reply button, it will be added to the correct comment? Do it using numbers? In addition, how do I keep track of the number of positive/negative replies for each comment?

Reply:
I made the above suggestion based on the fact that your comment doesn't have an id. Generally what I would prefer as a better solution on a high level is that, your Comment class contains a String id, with all the relevant data for the comment, including replies, and this will be the single source of truth.

On the front end side, you could have a method which creates a comment UI element from the Comment data, and its id will be used in the onClick listener. Each time the onClick listener triggered, you'll be talking to the server by providing the id, which can be used to identify the comment, as well as replies, because each reply can only belong to one comment, which is identified by the id.

For number of positive/negative replies, similarly, it is kept inside the Comment data. Your front end basically should not have much logic. It should only display according to data, and give server some updates through the unique id, and then update again when new data fetched.

Does this make sense?

Reply:
I see... It does! But I would like to constantly change the display (CSS shape) based on the sentiments. Does this mean that I should constantly perform queries? Personally, I feel that it will be a waste to perform the queries. Hence, I decided to weave the data component into the front-end. Thanks in advance! :D

Reply:
Hmmm how would you know about sentiment if you don't send it back to server? I guess you want to perform the sentiment analysis on front end?

What I would suggest is that the sentiment should be peroformed by the server, otherwise you may need to expose your api key on the front end which can be easily hacked in by others. So the flow would be when user replies, you would send the reply back to server, and wait for server to return you the sentiment of the reply, and then change the UI based on the result. Does this make sense?

Reply:
Oh no. I do want it to be done on the server side. However, I want the results to be saved on the front end. Specifically, for each comment, I would like to know that it has X number of positive replies and Y number of negative replies. In addition, I want to know whether the comment is positive/negative. 

In fact, I don't really need to store the replies. However, I did so in this PR because I don't want to complicate matters before I know that it works. 😅 

Nonetheless, thanks a lot for the advice! :D

What are your thoughts about the following?
- each comment will store its own sentiment, its id and the number of negative/positive replies
- a sentiment manager will use the sentiment data from the comments (I have been wanting to implement this manager but I wanted to make sure things work before going ahead with it)

I am not sure whether this will be how it pans out but I will try. 😂

Reply:
Yes the Comment looks good to me. I do see why things are getting a bit messy right now, but that is indeed a problem with vanila JS, so no worries if you find this not as trivial or elegant. If you later started to grab some web frametwork, like ReactJS or AngularJS, you'll find things much easier than this :)

Can you be more specific on what SentimentManager does?

Reply:
I see! HAHA. Yes. I have been wanting to learn a web framework (I did it very long ago but I didn't really appreciate then because I was building extremely basic websites). :D 

I have thought through the architecture in greater detail and decided that I will use the numerical values of the sentiments instead of positive/negative. Here is the architecture:

- `ColourUtilities` converts a sentiment value to a colour (via basic addition/multiplication)
- `SentimentColour` is an interface with `getColour` method
- `MultiSentimentColours` extends `Sentiment` and has `getResponseColour`(which returns an array of `SentimentColour`s) and `getId` methods
- `Reply` implements `SentimentColour` and has message, name and sentimentColour properties
- `Comment` implements `MultiSentiments` and has id, replies (an array of `Reply`s), sentimentColour, message and name properties 
- `Painter` takes in an array of `MultiSentiments` and computes the desired colour for each `MultiSentiment`

Basically, I would like to have the following data for my front-end (besides the name/message): a colour for each comment, a colour for each reply, and a colour for each comment + reply (i.e. the overall response). The latter colours are provided by `Painter`. 

Therefore, for each POST request, I will return a sentiment colour (for the comment/reply) and an array of colour values (which is provided by `Painter`). I will encapsulate these information using `SentimentAndPaints` so that I can easily serialise it into JSON format.

Reply:
A few things to notice:

1. How do you plan to have interface defined in javascript?
2. I think there's some misuse/misunderstanding about interface here. Interface should not be used when you found that two classes will have same methods, it should be used when you want different implementations of one object. For example, why would Reply implement SentimentColour? Reply itself is not some kind of color. Similarly, Comment is not a sentiment. The same things applies to MultiSentimentColours and Sentiment, how would a color extend a sentiment?

The main point is that we shouldn't over architect and force something that aren't relevant to be grouped together. Reply can have a color, Comment can have some colors, but they don't necessarily to be implementing an interface to have such a getColor/getResponseColour method, as we don't see a point where interface is important and concrete implementation is not. 

Think about List and ArrayList, the reason why we have List is that we don't want people to care about the implementations of the List, all they need to know is that a List can have these and these functionalities. Take Reply for example, we don't just care about the SentimentColour of it, we also care about its message, its name etc, same for Comment.

Based on this, do you have some neater design on it :)

Reply:
Thanks! Sorry, I just saw this.><  I am doing away with the interfaces. While implementing it, I realised that they are not necessary. I guess I was a tad too stuck with `Comment` and `Reply` when in fact I don't really need them -- besides for the conversion of JSON string.

Instead, I think what is more relevant is to create a `Palette` class that uses the data from each reply and comment. `Painter` will then store the `Palette`s. My reply and comment classes will just be used to create the formatted JSON string. 

On the front-end, I removed all the classes and I am only using function calls. 😅